### PR TITLE
ADAPT-1802: [hydra-publish] Adapt data classification changes according to our policy

### PR DESCRIPTION
**Pull request by aman-minz**
This PR introduces new Data Classification values in our system. As advised, we are also accepting the old values of data classification. However, the supported values/symbols for data classification in the v1 and v2 metadata schema are as per our policy. The old values are transformed to new values in DVS before they are sent to kafka-broker.

Please note that when an old value of data classification is received in the request, we take care of retaining it by storing in SubDataClassification field.

**DataClassification:**
`{
  "name": "dataClassification",
  "type": {
    "type": "enum",
    "name": "DataClassification",
    "namespace": "hydra.kafka.model",
    "symbols": [
      "Public",
      "InternalUse",
      "Confidential",
      "Restricted"
    ]
  }
}`

**SubDataClassification:**
`{
  "name": "subDataClassification",
  "type": [
    "null",
    {
      "type": "enum",
      "name": "SubDataClassification",
      "namespace": "hydra.kafka.model",
      "symbols": [
        "Public",
        "InternalUseOnly",
        "ConfidentialPII",
        "RestrictedFinancial",
        "RestrictedEmployeeData"
      ]
    }
  ],
  "default": null
}`

**Supported Values in request => Values sent to Kafka:**
1. Public                                 => Public
2. InternalUseOnly                => InternalUse
3. ConfidentialPII                  => Confidential
4. RestrictedFinancial          => Restricted
5. RestrictedEmployeeData => Restricted
6. InternalUse                       => InternalUse
7. Confidential                      => Confidential
8. Restricted                        => Restricted

### CODE REVIEW

File: ingestors/kafka/src/main/scala/hydra/kafka/programs/TopicMetadataError.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+sealed trait TopicMetadataError extends ValidationError
+
+object TopicMetadataError {
+
+  case class InvalidSubDataClassificationTypeError(dataClassification: String,
+                                                   subDataClassification: String,
+                                                   supportedValues: Seq[SubDataClassification]) extends TopicMetadataError {
+
+    override def message: String = {
+      val validValues = s"Valid value is ${if (supportedValues.size == 1) "'" + supportedValues.head + "'" else s"one of [${supportedValues.mkString(", ")}]"}."
```
Comment (shreedhar-kc) [inline ingestors/kafka/src/main/scala/hydra/kafka/programs/TopicMetadataError.scala:15]: Nit: 
I think + can be replaced here with s"'${supportedValues.head}'"

File: ingestors/kafka/src/main/scala/hydra/kafka/programs/TopicMetadataV2Validator.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  def validate(metadataV2Request: TopicMetadataV2Request): F[Unit] =
+    for {
+      _ <- validateSubDataClassification(metadataV2Request.dataClassification, metadataV2Request.subDataClassification)
+    } yield ()
+
+  private def validateSubDataClassification(dataClassification: DataClassification,
+                                            subDataClassification: Option[SubDataClassification]): F[Unit] = {
+    (subDataClassification map { sdc =>
+      val correctionRequired = validateSubDataClassificationValues(dataClassification, sdc)
```
Comment (shreedhar-kc) [inline ingestors/kafka/src/main/scala/hydra/kafka/programs/TopicMetadataV2Validator.scala]: This validation technique is nice.
But I feel like validateSubDataClassificationValues name is a bit misleading. The function is not really validating something but It looks like it is gathering records that needs correction or are incorrect.
Maybe changing the name to "collectInvalidSubDataClassifications()" or "gatherRequiredCorrections() or similar would improve readability.

**Review by shreedhar-kc — CHANGES_REQUESTED** [review]

File: ingestors/kafka/src/main/scala/hydra/kafka/programs/TopicMetadataV2Validator.scala (hunk 2 above)
Comment (aman-minz) [reply ingestors/kafka/src/main/scala/hydra/kafka/programs/TopicMetadataV2Validator.scala]: Sure, @shreedhar-kc. I have renamed it to `collectValidSubDataClassificationValues`. Is it fine?

**Review by shreedhar-kc — DISMISSED** [review]

**Review by abhivermaaa — DISMISSED** [review]
LGTM

**Review by aadit-chugh — DISMISSED** [review]

**Review by abhivermaaa — APPROVED** [review]
LGTM

**Review by shreedhar-kc — APPROVED** [review]

**Pull request merged**